From f08070f9ade1aa15bbc4844f14133e560ac3a532 Mon Sep 17 00:00:00 2001
From: Danny Lin <danny@kdrag0n.dev>
Date: Wed, 17 Mar 2021 16:29:59 -0700
Subject: [PATCH 43/61] SystemUI: Add machine learning back gesture provider

This is a provider for machine learning-assisted back gestures when
using gesture navigation. It uses TensorFlow Lite for inference, similar
to Google's Pixel implementation, so a model is necessary for it to
work.

Change-Id: I619837789f77d9a430a04afb68a8d8e68977e431
Signed-off-by: SamarV-121 <samarvispute121@gmail.com>
---
 packages/SystemUI/Android.bp                  |  3 ++
 packages/SystemUI/proguard.flags              |  3 +-
 .../CustomBackGestureTfClassifierProvider.kt  | 41 +++++++++++++++++++
 .../systemui/CustomSystemUIFactory.kt         | 10 +++++
 4 files changed, 56 insertions(+), 1 deletion(-)
 create mode 100644 packages/SystemUI/src/org/protonaosp/systemui/CustomBackGestureTfClassifierProvider.kt
 create mode 100644 packages/SystemUI/src/org/protonaosp/systemui/CustomSystemUIFactory.kt

diff --git a/packages/SystemUI/Android.bp b/packages/SystemUI/Android.bp
index 7e481112aac..f386604af94 100644
--- a/packages/SystemUI/Android.bp
+++ b/packages/SystemUI/Android.bp
@@ -74,6 +74,7 @@ android_library {
         "vendor.lineage.biometrics.fingerprint.inscreen-V1.0-java",
         "vendor.lineage.powershare-V1.0-java",
         "faceunlock_framework",
+        "tensorflowlite_java"
     ],
     manifest: "AndroidManifest.xml",
     additional_manifests: ["LineageManifest.xml"],
@@ -167,6 +168,8 @@ android_app {
     static_libs: [
         "SystemUI-core",
     ],
+    jni_libs: ["libtensorflowlite_jni"],
+    use_embedded_native_libs: true,
     resource_dirs: [],
 
     platform_apis: true,
diff --git a/packages/SystemUI/proguard.flags b/packages/SystemUI/proguard.flags
index 2a2ba1b0cca..e6346489664 100644
--- a/packages/SystemUI/proguard.flags
+++ b/packages/SystemUI/proguard.flags
@@ -16,6 +16,7 @@
 -keep class com.android.systemui.statusbar.tv.TvStatusBar
 -keep class com.android.systemui.car.CarSystemUIFactory
 -keep class com.android.systemui.SystemUIFactory
+-keep class org.protonaosp.systemui.CustomSystemUIFactory
 -keep class com.android.systemui.tv.TvSystemUIFactory
 -keep class * extends com.android.systemui.SystemUI
 -keep class * implements com.android.systemui.SystemUI$Injector
@@ -39,4 +40,4 @@
 
 -keep public class * extends com.android.systemui.SystemUI {
     public <init>(android.content.Context);
-}
\ No newline at end of file
+}
diff --git a/packages/SystemUI/src/org/protonaosp/systemui/CustomBackGestureTfClassifierProvider.kt b/packages/SystemUI/src/org/protonaosp/systemui/CustomBackGestureTfClassifierProvider.kt
new file mode 100644
index 00000000000..dee37d99dd5
--- /dev/null
+++ b/packages/SystemUI/src/org/protonaosp/systemui/CustomBackGestureTfClassifierProvider.kt
@@ -0,0 +1,41 @@
+package org.protonaosp.systemui
+
+import android.content.res.AssetManager
+import com.android.systemui.statusbar.phone.BackGestureTfClassifierProvider
+import org.tensorflow.lite.Interpreter
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+
+class CustomBackGestureTfClassifierProvider(
+    am: AssetManager,
+    private val modelName: String
+) : BackGestureTfClassifierProvider() {
+    // Don't bother to set up a MappedByteBuffer for 512 KiB of data
+    private val interpreter = am.open("$modelName.tflite").use {
+        val data = it.readBytes()
+        Interpreter(ByteBuffer.allocateDirect(data.size).apply {
+            order(ByteOrder.nativeOrder())
+            put(data)
+        })
+    }
+
+    override fun loadVocab(am: AssetManager): Map<String, Int> {
+        am.open("$modelName.vocab").use { ins ->
+            return String(ins.readBytes()).lines().asSequence()
+                .withIndex()
+                .map { it.value to it.index }
+                .toMap()
+        }
+    }
+
+    override fun predict(featuresVector: Array<Any>): Float {
+        val confidenceTensor = floatArrayOf(0f)
+        val tensors = mutableMapOf(0 to arrayOf(confidenceTensor))
+
+        interpreter.runForMultipleInputsOutputs(featuresVector, tensors as Map<Int, Any>)
+        return confidenceTensor[0]
+    }
+
+    override fun release() = interpreter.close()
+    override fun isActive() = true
+}
diff --git a/packages/SystemUI/src/org/protonaosp/systemui/CustomSystemUIFactory.kt b/packages/SystemUI/src/org/protonaosp/systemui/CustomSystemUIFactory.kt
new file mode 100644
index 00000000000..66e883ec313
--- /dev/null
+++ b/packages/SystemUI/src/org/protonaosp/systemui/CustomSystemUIFactory.kt
@@ -0,0 +1,10 @@
+package org.protonaosp.systemui
+
+import android.content.res.AssetManager
+import com.android.systemui.SystemUIFactory
+
+class CustomSystemUIFactory : SystemUIFactory() {
+    // ML back gesture provider
+    override fun createBackGestureTfClassifierProvider(am: AssetManager, modelName: String) =
+        CustomBackGestureTfClassifierProvider(am, modelName)
+}
-- 
2.32.0

